riverdriver/riverpgxv5/river_pgx_v5_driver_test.go

```diff
@@ -34,10 +39,144 @@ func TestNew(t *testing.T) {
 	})
 }
 
+func TestListener_Close(t *testing.T) {
+	t.Parallel()
+
+	ctx := context.Background()
+
+	t.Run("NoOpWithoutConn", func(t *testing.T) {
+		t.Parallel()
+
+		listener := &Listener{dbPool: testPool(ctx, t, nil)}
+		require.Nil(t, listener.conn)
+		require.NoError(t, listener.Close(ctx))
+	})
+
+	t.Run("ReleasesAndUnsetsConn", func(t *testing.T) {
+		t.Parallel()
+
+		config := testPoolConfig()
+
+		releaseInvoked := make(chan struct{})
+
+		// pgx calls AfterRelease in a goroutine, which is why we communicate with a channel.
+		config.AfterRelease = func(c *pgx.Conn) bool {
+			close(releaseInvoked)
+			return true
+		}
+
+		listener := &Listener{dbPool: testPool(ctx, t, config)}
+
+		require.NoError(t, listener.Connect(ctx))
+		require.NotNil(t, listener.conn)
+
+		require.NoError(t, listener.Close(ctx))
+
+		// Connection has been released.
+		select {
+		case <-releaseInvoked:
+		case <-time.After(3 * time.Second):
+			require.FailNow(t, "Timed out waiting for connection to be released")
+		}
+
+		// Despite error, internal connection still unset.
+		require.Nil(t, listener.conn)
+	})
+
+	t.Run("UnsetsConnEvenOnError", func(t *testing.T) {
+		t.Parallel()
+
+		var connStub *connStub
+
+		config := testPoolConfig()
+		config.ConnConfig.DialFunc = func(ctx context.Context, network, addr string) (net.Conn, error) {
+			conn, err := (&net.Dialer{KeepAlive: 5 * time.Minute}).Dial(network, addr)
+			if err != nil {
+				return nil, err
+			}
+
+			connStub = newConnStub(conn)
+			return connStub, nil
+		}
+
+		listener := &Listener{dbPool: testPool(ctx, t, config)}
+
+		require.NoError(t, listener.Connect(ctx))
+		require.NotNil(t, listener.conn)
+
+		conn := listener.conn.Conn()
+
+		expectedErr := errors.New("conn close error")
+		connStub.closeFunc = func() error {
+			t.Logf("Close invoked; returning error")
+			return expectedErr
+		}
+
+		// Error isn't passed through because we invoke pgxpool.Release.
+		require.NoError(t, listener.Close(ctx))
+
+		// Despite error, internal connection still unset.
+		require.Nil(t, listener.conn)
+
+		// Make sure that our stubbing above worked by closing the connection we
+		// retained and verifying it's the error we set.
+		require.ErrorIs(t, conn.Close(ctx), expectedErr)
+	})
+}
+
 func TestInterpretError(t *testing.T) {
 	t.Parallel()
 
 	require.EqualError(t, interpretError(errors.New("an error")), "an error")
 	require.ErrorIs(t, interpretError(pgx.ErrNoRows), rivertype.ErrNotFound)
 	require.NoError(t, interpretError(nil))
 }
+
+// connStub implements net.Conn and allows us to stub particular functions like
+// Close that are otherwise nigh impossible to test.
+type connStub struct {
+	net.Conn
+
+	closeFunc func() error
+}
+
+func newConnStub(conn net.Conn) *connStub {
+	return &connStub{
+		Conn: conn,
+
+		closeFunc: conn.Close,
+	}
+}
+
+func (c *connStub) Close() error {
+	return c.closeFunc()
+}
+
+func testPoolConfig() *pgxpool.Config {
```

Comment:
Maybe `riverinternaltest` should be its own package so we can reuse this stuff, but ugh, skipped any further refactoring there for now.

Reply:
not more packages 😩

Reply:
Haha, ack.